src/box/txn_limbo.c

```diff
@@ -208,7 +212,16 @@ txn_limbo_abort(struct txn_limbo *limbo, struct txn_limbo_entry *entry)
 	 * rollback in the reversed order. Rolled back transaction
 	 * is always the last.
 	 */
+	if (entry == limbo->confirmed_entry) {
+		limbo->confirmed_entry = rlist_prev_entry(
+			limbo->confirmed_entry, in_queue);
+		limbo->confirmed_entry_is_valid = false;
+	} else if (entry == rlist_next_entry(limbo->confirmed_entry,
+					     in_queue)) {
+		limbo->confirmed_entry_is_valid = false;
+	}
```

Comment:
- How could a confirmed entry be aborted? After a txn is confirmed (its lsn <= confirmed_lsn), it is 100% commited, hm?
- Why does it become invalid? A txn was aborted, ok. If somehow the aborted txn was the confirmed one, you make a step back and it is valid again - confirm of LSN L automatically means all LSNs < L are also confirmed. Otherwise if the aborted txn was after the confirmed LSN, there seem to be no reason why it would affect the confirmed entry. Hm?

Reply:
- I was wrong in that comment about the invariant (fixed). There should be an inverse inequality. `confirmed_lsn <= confirmed_entry->lsn`. Therefore, in the case where `confirmed_lsn < confirmed_entry->lsn` this may well happen.
- > If somehow the aborted txn was the confirmed one.

But this cannot be. Aborted txn can be `confirmed_entry`, but not a confirmed transaction. Yes, this is bad naming, as I already said. `confirmed_entry` can point to an asynchronous transaction located after the last **actually** confirmed synchronous transaction. Let's rename it something so no one gets confused.
> Otherwise if the aborted txn was after the confirmed LSN, there seem to be no reason why it would affect the confirmed entry.

But we have an invariant `confirmed_entry_is_valid == true` <=> the next transaction after `confirmed_entry` is synchronous and has lsn (not -1) and `ack_count` stores the correct value. Here it breaks down.

Reply:
One more comment about this invariant `confirmed_lsn <= confirmed_entry->lsn`: 
I just found out that it is only valid if the following patch is applied:
```diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index 458b3027f..97f50886e 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -708,13 +718,17 @@ txn_limbo_confirm(struct txn_limbo *limbo)
        assert(e->lsn != -1);
        int32_t k = (int32_t)vclock_size(&limbo->vclock)
                - replication_synchro_quorum;
-       int64_t confirm_lsn = (k < 0) ? 0 :
+       int64_t confirm_lsn1 = (k < 0) ? 0 :
                vclock_nth_element(&limbo->vclock, k);
-       assert(confirm_lsn != -1);
+       assert(confirm_lsn1 != -1);
+       int64_t confirm_lsn2 = -1;
        for (; !rlist_entry_is_head(e, &limbo->queue, in_queue);
-              e = rlist_next_entry(e, in_queue))
-               if (e->lsn == -1 || e->lsn > confirm_lsn)
+              e = rlist_next_entry(e, in_queue)) {
+               if (e->lsn == -1 || e->lsn > confirm_lsn1)
                        break;
+               if (txn_has_flag(e->txn, TXN_WAIT_ACK))
+                       confirm_lsn2 = e->lsn;
+       }
        limbo->confirmed_entry = rlist_prev_entry(e, in_queue);
        /*
         * We have not made sure that the transaction following confirmed_entry
@@ -722,15 +736,15 @@ txn_limbo_confirm(struct txn_limbo *limbo)
         * valid.
         */
        limbo->confirmed_entry_is_valid = false;
-       txn_limbo_write_confirm(limbo, confirm_lsn);
-       txn_limbo_read_confirm(limbo, confirm_lsn);
+       txn_limbo_write_confirm(limbo, confirm_lsn2);
+       txn_limbo_read_confirm(limbo, confirm_lsn2);
 }
 ```

Reply:
Hm. Yeah, ok, the naming breaks my brain 😵‍💫. This invariant is very hard to keep in the head when it is not about confirmed entry itself, but about the next one. Have you considered what happens if you store the pointer at that next entry instead? It would be less cognitive load if the entry and its guarantees would be one. Not like entry is one, and the guarantees are for another entry related to the first one somehow.

Reply:
You can think of it this way:
That is, it’s just `confirmed_entry` is a transaction such that all previous ones are committed. You can probably call it `committed_entry` or `last_committed_entry` (by committed I mean that it can no longer be rolled back). This fully expresses the meaning of this variable.